src/Farmer/Builders/Builders.VirtualNetwork.fs

```diff
@@ -409,6 +419,74 @@ type VNetPeeringSpecBuilder() =
 
 let vnetPeering = VNetPeeringSpecBuilder ()
 
+type PrivateLinkServiceConnection = 
+    { Resource: LinkedResource 
+      GroupIds: string list }
+
+type PrivateEndpointConfig = 
+    { Name: ResourceName
+      Subnet : SubnetReference option
+      PrivateLinkServiceConnection: PrivateLinkServiceConnection option
+      PrivateDnsZone: LinkedResource option }
+    interface IBuilder with
+        member this.ResourceId = privateEndpoints.resourceId this.Name
+        member this.BuildResources location = 
+            let serviceConn = 
+                match this.PrivateLinkServiceConnection with
+                | Some sc -> sc
+                | None -> raiseFarmer "Private endpoint must be attached to a resource"
+
+            let endpoint = 
+                { Name = this.Name
+                  Subnet = match this.Subnet with | Some sn -> sn | None -> raiseFarmer "Must have linked subnet"
+                  Location = location
+                  Resource = serviceConn.Resource
+                  GroupIds = serviceConn.GroupIds }
+
+            [
+                endpoint
+
+                match this.PrivateDnsZone with 
+                | None -> ()
+                | Some zone -> {
+                    Name = (ResourceName $"{this.Name.Value}/dnszone")
+                    Location = location
+                    PrivateEndpoint = (Managed (this :> IBuilder).ResourceId)
+                    PrivateDnsZone = zone }
+            ]
+
+type PrivateEndpointBuilder() =
+    member _.Yield _ =
+        { Name = ResourceName.Empty
+          Subnet = None
+          PrivateLinkServiceConnection = None
+          PrivateDnsZone = None }
+    
+    [<CustomOperation "name">]
+    member _.Name(state:PrivateEndpointConfig, name:string) = { state with Name = (ResourceName name) }
+    member _.Name(state:PrivateEndpointConfig, name:ResourceName) = { state with Name = name }
+
+    [<CustomOperation "subnet">]
+    member _.Subnet(state:PrivateEndpointConfig, subnet:SubnetReference) = { state with Subnet = Some subnet }
+    member _.Subnet(state:PrivateEndpointConfig, subnet:SubnetConfig) = { state with Subnet = Some (SubnetReference.Direct (Managed (subnet :> IBuilder).ResourceId)) }
```

Comment:
If you moved this below the type extension on Lines 490-492, you could use `SubnetReference.create(subnet)` to simplify this expression